packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,149 @@
+import { SortType } from 'design/DataTable/types';
+import { useAsync } from 'shared/hooks/useAsync';
+import { ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useClusterContext } from '../clusterContext';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
+import { ResourceKind } from 'teleport/Discover/Shared';
+
+export function useServerSideResources<Agent>(
+  fetchFunction: (params: ServerSideParams) => Promise<{
+    agentsList: Array<Agent>;
+    totalCount: number;
+    startKey: string;
+  }>
+) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: getDefaultSort(),
+  });
+  const [pageCount, setPageCount] = useState({
+    to: 0,
+    from: 0,
+    total: 0,
+  });
+
+  const limit = 15;
+
+  // startKey is used here as a way to paginate through agents returned from
+  // their respective rpcs.
+  const [fetchAttempt, fetch] = useAsync(async (startKey: string) =>
+    retryWithRelogin(ctx, documentUri, clusterUri, () =>
+      fetchFunction({
+        ...agentFilter,
+        limit,
+        clusterUri,
+        startKey,
+      })
+    )
+  );
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        fetch(proposedKey);
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        // if we got pageIndex - 1, we'd get the same data as the previous "next" we just called.
+        // we have to go back 2. which is why we use newPageIndex instead
+
+        // example after going from pageIndex = 0 to pageIndex = 1:
+        // ['startKeyFromInitialFetch', 'goEvenFurtherStartKey']
+        //     ^- just came from here          ^- we are here
+        //
+        // so in order to get the start key for the page BEFORE this one, we go back twice.
+        // this could also be written/read as keys[pageIndex - 2]
+        fetch(keys[newPageIndex - 1]);
+        setPageIndex(newPageIndex);
+      };
+    }
+    return null;
+  }
+
+  useEffect(() => {
+    // when agent filter changes we consider it a 'new' search so we should pass an empty
+    // startKey to get the beginning of the list with new filters, and set pageIndex back to zero
+    fetch('');
+    setPageIndex(0);
+  }, [agentFilter]);
```

Comment:
If the page index needs to be set to 0 when the agent filter changes, we should do so in functions which update the agent filter.

See:

* https://beta.reactjs.org/learn/you-might-not-need-an-effect#chains-of-computations
* https://beta.reactjs.org/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers

Reply:
Updated 👍